README.md
byte
=======

[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![Test coverage][coveralls-image]][coveralls-url]
[![Gittip][gittip-image]][gittip-url]
[![David deps][david-image]][david-url]
[![node version][node-image]][node-url]
[![npm download][download-image]][download-url]

[npm-image]: https://img.shields.io/npm/v/byte.svg?style=flat-square
[npm-url]: https://npmjs.org/package/byte
[travis-image]: https://img.shields.io/travis/node-modules/byte.svg?style=flat-square
[travis-url]: https://travis-ci.org/node-modules/byte
[coveralls-image]: https://img.shields.io/coveralls/node-modules/byte.svg?style=flat-square
[coveralls-url]: https://coveralls.io/r/node-modules/byte?branch=master
[gittip-image]: https://img.shields.io/gittip/fengmk2.svg?style=flat-square
[gittip-url]: https://www.gittip.com/fengmk2/
[david-image]: https://img.shields.io/david/node-modules/byte.svg?style=flat-square
[david-url]: https://david-dm.org/node-modules/byte
[node-image]: https://img.shields.io/badge/node.js-%3E=_0.10-green.svg?style=flat-square
[node-url]: http://nodejs.org/download/
[download-image]: https://img.shields.io/npm/dm/byte.svg?style=flat-square
[download-url]: https://npmjs.org/package/byte

![logo](https://raw.github.com/node-modules/byte/master/logo.png)

Input Buffer and Output Buffer,
just like Java [`ByteBuffer`](http://docs.oracle.com/javase/6/docs/api/java/nio/ByteBuffer.html).

## Install

```bash
$ npm install byte --save
```

## Usage

All methods just like Java ByteBuffer,
you find them out [here](http://docs.oracle.com/javase/6/docs/api/java/nio/ByteBuffer.html#method_summary).

```js
var ByteBuffer = require('byte');

var bb = ByteBuffer.allocate(1024);
bb.order(ByteBuffer.BIG_ENDIAN); // default is BIG_ENDIAN, you can change it to LITTLE_ENDIAN.
bb.put(0);
bb.put(new Buffer([0, 1, 2]));
bb.put(new Buffer([255, 255, 255, 255]), 10, 3);
bb.put(21, 100);

bb.putChar('a');
bb.putChar(10, 'b');

bb.putInt(1024);
bb.putInt(-100);

bb.putFloat(100.9);

bb.putLong(10000100009099);
bb.putLong('1152921504606847000');

bb.putShort(65535);
bb.putShort(-50000);

bb.putDouble(99.99999);

// wrap for read
var rb = ByteBuffer.wrap(new Buffer(100));
rb.getInt();
rb.getLong();
rb.getChar();
rb.get();
rb.getDouble();
rb.getFloat();
```

## Benchmark

```bash
$ node benchmark/put.js

node version: v0.11.12, date: Mon May 12 2014 18:25:35 GMT+0800 (CST)
Starting...
20 tests completed.

put()                                      x 29,971,599 ops/sec ±4.10% (96 runs sampled)
putChar("a")                               x 27,950,189 ops/sec ±6.22% (80 runs sampled)
putChar(61)                                x 34,798,492 ops/sec ±5.08% (81 runs sampled)
putShort()                                 x 25,264,781 ops/sec ±2.90% (88 runs sampled)
putInt()                                   x 21,368,588 ops/sec ±6.07% (85 runs sampled)
putFloat()                                 x 12,324,148 ops/sec ±2.04% (93 runs sampled)
putDouble()                                x 13,374,686 ops/sec ±1.41% (92 runs sampled)
putLong(100000)                            x 17,754,878 ops/sec ±5.16% (86 runs sampled)
putSmallSLong("10000")                     x  7,732,989 ops/sec ±2.07% (92 runs sampled)
putBigNumLong(34359738368)                 x  3,580,231 ops/sec ±2.58% (93 runs sampled)
putSafeStrLong("34359738368")              x  2,443,560 ops/sec ±2.04% (97 runs sampled)
putStrLong("9223372036854775808")          x    760,908 ops/sec ±2.42% (92 runs sampled)
ByteBuffer.allocate(100).putString(0, str) x    608,403 ops/sec ±11.46% (70 runs sampled)
putString(0, str)                          x  1,362,412 ops/sec ±8.55% (85 runs sampled)
bytes.putString(str)                       x  1,506,610 ops/sec ±2.31% (94 runs sampled)
putString(0, buf)                          x  5,947,594 ops/sec ±4.16% (90 runs sampled)
bytes.putString(buf)                       x  5,741,251 ops/sec ±1.69% (95 runs sampled)
putRawString(0, str)                       x  2,908,161 ops/sec ±1.81% (95 runs sampled)
bytes.putRawString(str)                    x  1,527,089 ops/sec ±4.98% (86 runs sampled)
bytes.putRawString(str).array()            x  1,009,026 ops/sec ±2.38% (91 runs sampled)

$node benchmark/get.js

node version: v0.11.12, date: Mon May 12 2014 19:14:26 GMT+0800 (CST)
Starting...
15 tests completed.

get(0, 1) => copy Buffer    x  2,059,464 ops/sec ±9.18% (69 runs sampled)
get(0, 100) => copy Buffer  x  2,124,455 ops/sec ±4.98% (75 runs sampled)
get(0, 4096) => copy Buffer x    356,927 ops/sec ±9.43% (56 runs sampled)
get() => byte               x 15,477,897 ops/sec ±3.05% (89 runs sampled)
getChar(0)                  x 52,541,591 ops/sec ±1.04% (95 runs sampled)
getShort(0)                 x 26,297,086 ops/sec ±2.46% (89 runs sampled)
getInt(0)                   x 18,772,003 ops/sec ±6.27% (71 runs sampled)
getFloat(0)                 x 13,132,298 ops/sec ±1.68% (97 runs sampled)
getDouble(0)                x 10,968,594 ops/sec ±1.27% (94 runs sampled)
getLong(0)                  x 11,849,374 ops/sec ±2.63% (96 runs sampled)
getString(0)                x  2,358,382 ops/sec ±5.78% (76 runs sampled)
getCString(0)               x  1,618,356 ops/sec ±8.41% (72 runs sampled)
readRawString(4, 100)       x  4,790,991 ops/sec ±9.25% (79 runs sampled)
readRawString(100)          x  5,434,663 ops/sec ±1.32% (95 runs sampled)
getRawString(0, 100)        x  5,497,325 ops/sec ±1.02% (98 runs sampled)
```

## `Number` methods

```
putShort / putInt16
putUInt16
putInt / putInt32
putUInt / putUInt32
putInt64
putFloat
putDouble


getShort / getInt16
getUInt16
getInt / getInt32
getUInt / getUInt32
getInt64
getFloat
getDouble
```

## `String` methods

Java String format: `| length (4 bytes int) | string bytes |`

C String format: `| length + 1 (4 bytes int) | string bytes | \0 |`

Row String format: `string bytes`

### `putString()` and `putCString()` and `putRawString()`

```js
bb.putString('foo');
bb.putString(new Buffer('foo'));

bb.putCString('foo');
bb.putCString(new Buffer('foo'));

bb.putRawString('foo');
```

### `getString()` and `getCString()` and `getRawString(), readRawString()`

```js
bb.getString();
bb.getString(10);

bb.getCString();
bb.getCString(10);

bb.getRawString(0, 10);
bb.readRawString(10);
```

## License

(The MIT License)

Copyright (c) 2013 - 2014 fengmk2 &lt;fengmk2@gmail.com&gt; and other contributors

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md

1.4.1 / 2018-05-15
==================

**fixes**
  * [[`ccc6f4b`](http://github.com/node-modules/byte/commit/ccc6f4b206d01761c38ca2717014b646ce8f4c93)] - fix: using Buffer.alloc(size) instead of new Buffer(size) (#34) (zōng yǔ <<gxcsoccer@users.noreply.github.com>>)

1.4.0 / 2018-01-30
==================

**features**
  * [[`2f29793`](http://github.com/node-modules/byte/commit/2f2979376256dd7deae4cc312c4dd019e1ab13ea)] - feat: add memcpy() (#32) (Khaidi Chu <<i@2333.moe>>)

1.3.0 / 2017-11-16
==================

**features**
  * [[`8dd202f`](http://github.com/node-modules/byte/commit/8dd202f06fba076169b9ff3231e65cfc21140555)] - feat: support get raw string by string length (#30) (Yiyu He <<dead_horse@qq.com>>)

**others**
  * [[`7dd0a0c`](http://github.com/node-modules/byte/commit/7dd0a0c7dda5a0c3e5bf49b3449767e69dfc43b5)] - chore: remove duplicated commits on History.md (fengmk2 <<fengmk2@gmail.com>>)

1.2.0 / 2017-08-04
==================

**features**
  * [[`2a052c5`](http://github.com/node-modules/byte/commit/2a052c55c065120a508c98e4595bc11f0c96275a)] - feat: improve putRawString performance (#27) (fengmk2 <<fengmk2@gmail.com>>)

1.1.6 / 2017-05-11
==================

  * refactor: enhance ByteBuffer.get performance

1.1.5 / 2015-09-28
==================

  * fix: getRawString emoji

1.1.4 / 2015-07-21
==================

 * Fix the bug of allocating insufficient space (@pmq20)
 * Invoke _checkSize when index is provided

1.1.3 / 2015-07-01
==================

 * fix: putRawString emoji

1.1.2 / 2015-05-22
==================

 * deps: upgrade debug, long, utility

1.1.1 / 2014-10-23
==================

 * fix bugs: buffer copy (@gxcsoccer)

1.1.0 / 2014-10-18
==================

 * add missing methods (@gxcsoccer)

1.0.0 / 2014-08-27
==================

 * add optimized for numbers
 * refact put
 * refactor putChar
 * optimize put, putChar
 * add 0.11.13 benchmark result
 * add test ubffer slice and copy in node 0.10.28
 * add write int and fill byte benchmark
 * add more benchmarks

0.3.1 / 2014-05-15
==================

  * Merge pull request #9 from node-modules/array-as-copy
  * exports copy bytes, support copy(start, end)

0.3.0 / 2014-05-13
==================

 * add more buffer copy test cases
 * array() must always copy. close #8
 * improve getString performance
 * return this on putNumber*
 * fix putLong convert
 * support auto change save string long to number
 * add test for putString
 * add test for put buffer string
 * improve getString, getRawString
 * improve putString
 * add putBufString benchmark
 * improve putRawString
 * improve performance
 * use ~ for deps
 * fix jshint
 * fix benchmark string
 * add test for fastSlice
 * add benchmark for string, use _fastSlice
 * remove dynamic method name

0.2.3 / 2014-05-07 
==================

  * remove annoying warn

0.2.2 / 2014-04-22
==================

 * use taobao npm
 * use instanbul and jshint
 * use n.high and n.low to check Long instance, because long module will be diff in the top project

0.2.1 / 2014-02-08 
==================

  * Support (U)Int8 8 bit int

0.2.0 / 2014-01-25 
==================

  * add test for rawString (@dead-horse)
  * add rawString methods (@dead-horse)

0.1.3 / 2014-01-23 
==================

  * update authors
  * add test for uint, uint16
  * refactor put and get number of bytes (@dead-horse)
  * install from cnpm

0.1.2 / 2013-11-14 
==================

  * add read(size)

0.1.1 / 2013-11-14 
==================

  * handle empty string

0.1.0 / 2013-11-14 
==================

  * add getString() and putString()
  * add ByteBuffer.wrap()
  * add ByteBuffer get*()

0.0.2 / 2013-11-13 
==================

  * add warnning when buffer reallocate

0.0.1 / 2013-11-13 
==================

  * add benchmark
  * add putInt(), putChar(), putDouble(), put*()
  * update readme
  * first commit


benchmark/put.js
/**!
 * byte - benchmark/put.js
 *
 * Copyright(c) fengmk2 and other contributors.
 * MIT Licensed
 *
 * Authors:
 *   fengmk2 <fengmk2@gmail.com> (http://fengmk2.github.com)
 *   dead-horse <dead_horse@qq.com> (https://github.com/dead-horse)
 */

"use strict";

/**
 * Module dependencies.
 */

var Benchmark = require('benchmark');
var benchmarks = require('beautify-benchmark');
var ByteBuffer = require('../');

var suite = new Benchmark.Suite();

var putBytes = ByteBuffer.allocate(1);
var putCharBytes = ByteBuffer.allocate(1);
var putCharNumBytes = ByteBuffer.allocate(1);
var putShortBytes = ByteBuffer.allocate(2);
var putIntBytes = ByteBuffer.allocate(4);
var putFloatBytes = ByteBuffer.allocate(4);
var putDoubleBytes = ByteBuffer.allocate(8);
var putLongBytes = ByteBuffer.allocate(8);
var putBigLongBytes = ByteBuffer.allocate(8);
var putStringBytes = ByteBuffer.allocate(100);
var putBufStringBytes = ByteBuffer.allocate(100);
var putRawStringBytes = ByteBuffer.allocate(100);

var bytes = ByteBuffer.allocate(1024 * 1024);

var string = '';
while(string.length < 100) {
  string += 'fffff';
}
var bufString = new Buffer(100);

var buf = new Buffer(10);
var src = new Buffer([0x61, 0x62, 0x63]);

suite
.add('buf[0] = 97', function () {
  buf[0] = 97;
})
.add('put(0, 97)', function () {
  putBytes.put(0, 97);
})
.add('copy[0] = 97', function () {
  src.copy(buf, 0, 0, 1);
})
.add('put(<Buffer 61>)', function () {
  putBytes.put(src, 0, 1);
})
.add('putChar(0, "a")', function () {
  putCharBytes.putChar(0, 'a');
})
.add('putChar(0, 97)', function () {
  putCharNumBytes.putChar(0, 97);
})
.add('putChar("a")', function () {
  putCharBytes._offset = 0;
  putCharBytes.putChar('a');
})
.add('putChar(97)', function () {
  putCharNumBytes._offset = 0;
  putCharNumBytes.putChar(97);
})
.add('buf.writeUInt16BE(0, 1)', function () {
  buf.writeUInt16BE(1, 0);
})
.add('buf[uint16BE](0, 1)', function () {
  var uint16BE = 'writeUInt16BE';
  buf[uint16BE](1, 0);
})
.add('putShort(0, 1)', function () {
  putShortBytes.putShort(0, 1);
})
.add('putInt(0, 1)', function () {
  putIntBytes.putInt(0, 1);
})
.add('putFloat(0, 1)', function () {
  putFloatBytes.putFloat(0, 1);
})
.add('putDouble(0, 1)', function () {
  putDoubleBytes.putDouble(0, 1);
})
.add('putLong(0, 100000)', function () {
  putLongBytes.putLong(0, 100000);
})
.add('putSmallSLong(0, "10000") ', function () {
  putLongBytes.putLong(0, '10000');
})
.add('putBigNumLong(0, 34359738368) ', function () {
  putLongBytes.putLong(0, 34359738368);
})
.add('putSafeStrLong(0, "34359738368")', function () {
  putLongBytes.putLong(0, "34359738368");
})
.add('putStrLong(0, "9223372036854775808")', function () {
  putLongBytes.putLong(0, '9223372036854775808');
})
.add('ByteBuffer.allocate(100).putString(0, str)', function () {
  ByteBuffer.allocate(100).putString(0, string);
})
.add('putString(0, str)', function () {
  putStringBytes.putString(0, string);
})
.add('bytes.putString(str)', function () {
  bytes.reset();
  bytes.putString(string);
})
.add('putString(0, buf)', function () {
  putBufStringBytes.putString(0, bufString);
})
.add('bytes.putString(buf)', function () {
  bytes.reset();
  bytes.putString(bufString);
})
.add('putRawString(0, str)', function () {
  putRawStringBytes.putRawString(0, string);
})
.add('bytes.putRawString(str)', function () {
  bytes.reset();
  bytes.putRawString(string);
})
.add('bytes.putRawString(str).array()', function () {
  bytes.reset();
  bytes.putRawString(string).array();
})
.add('bytes.putRawString(str).array(0, 100)', function () {
  bytes.reset();
  bytes.putRawString(string).array(0, 100);
})

.on('cycle', function(event) {
  benchmarks.add(event.target);
})
.on('start', function(event) {
  console.log('\n  node version: %s, date: %s\n  Starting...', process.version, Date());
})
.on('complete', function done() {
  benchmarks.log();
})
.run({ 'async': false });

  // node version: v0.11.13, date: Mon Aug 25 2014 20:53:00 GMT+0800 (CST)

  // buf[0] = 97                                x 75,501,118 ops/sec ±2.29% (89 runs sampled)
  // put(0, 97)                                 x 76,489,133 ops/sec ±1.68% (93 runs sampled)
  // copy[0] = 97                               x 11,259,189 ops/sec ±2.01% (91 runs sampled)
  // put(<Buffer 61>)                           x 10,826,427 ops/sec ±1.68% (92 runs sampled)
  // putChar(0, "a")                            x 77,247,378 ops/sec ±1.74% (94 runs sampled)
  // putChar(0, 97)                             x 70,058,699 ops/sec ±2.18% (92 runs sampled)
  // putChar("a")                               x 63,099,885 ops/sec ±0.75% (94 runs sampled)
  // putChar(97)                                x 62,164,356 ops/sec ±1.38% (95 runs sampled)
  // putShort(0, 1)                             x 19,918,200 ops/sec ±1.03% (95 runs sampled)
  // putInt(0, 1)                               x  7,027,621 ops/sec ±6.70% (91 runs sampled)
  // putFloat(0, 1)                             x 10,104,306 ops/sec ±1.04% (91 runs sampled)
  // putDouble(0, 1)                            x 10,157,512 ops/sec ±1.10% (95 runs sampled)
  // putLong(0, 100000)                         x 26,364,254 ops/sec ±0.94% (93 runs sampled)
  // putSmallSLong(0, "10000")                  x 13,080,481 ops/sec ±1.18% (91 runs sampled)
  // putBigNumLong(0, 34359738368)              x  7,608,611 ops/sec ±1.17% (90 runs sampled)
  // putSafeStrLong(0, "34359738368")           x  4,070,055 ops/sec ±0.88% (93 runs sampled)
  // putStrLong(0, "9223372036854775808")       x    985,469 ops/sec ±0.89% (95 runs sampled)
  // ByteBuffer.allocate(100).putString(0, str) x    149,300 ops/sec ±1.60% (86 runs sampled)
  // putString(0, str)                          x    914,022 ops/sec ±3.70% (90 runs sampled)
  // bytes.putString(str)                       x    973,302 ops/sec ±0.72% (97 runs sampled)
  // putString(0, buf)                          x  2,535,609 ops/sec ±0.93% (94 runs sampled)
  // bytes.putString(buf)                       x  2,570,024 ops/sec ±3.20% (97 runs sampled)
  // putRawString(0, str)                       x  2,514,063 ops/sec ±0.74% (98 runs sampled)
  // bytes.putRawString(str)                    x  1,479,018 ops/sec ±0.76% (96 runs sampled)
  // bytes.putRawString(str).array()            x    325,139 ops/sec ±1.71% (83 runs sampled)
  // bytes.putRawString(str).array(0, 100)      x    333,073 ops/sec ±1.72% (85 runs sampled)


  // node version: v0.10.31, date: Mon Aug 25 2014 22:21:22 GMT+0800 (CST)
  // Starting...

  // buf[0] = 97      x 81,416,619 ops/sec ±2.65% (90 runs sampled)
  // put(0, 97)       x 39,789,758 ops/sec ±1.48% (92 runs sampled)
  // copy[0] = 97     x  7,876,033 ops/sec ±2.53% (96 runs sampled)
  // put(<Buffer 61>) x  6,758,320 ops/sec ±1.45% (91 runs sampled)
  // putChar(0, "a")  x 42,881,892 ops/sec ±0.60% (92 runs sampled)
  // putChar(0, 97)   x 41,615,714 ops/sec ±2.85% (75 runs sampled)
  // putChar("a")     x 33,757,160 ops/sec ±1.78% (92 runs sampled)
  // putChar(97)      x 36,055,373 ops/sec ±0.85% (97 runs sampled)

  // node version: v0.11.12, date: Mon Aug 25 2014 22:23:09 GMT+0800 (CST)
  // Starting...

  // buf[0] = 97      x 83,861,096 ops/sec ±1.38% (91 runs sampled)
  // put(0, 97)       x 46,275,837 ops/sec ±1.50% (94 runs sampled)
  // copy[0] = 97     x 11,822,231 ops/sec ±2.10% (91 runs sampled)
  // put(<Buffer 61>) x 10,097,220 ops/sec ±2.56% (87 runs sampled)
  // putChar(0, "a")  x 47,841,231 ops/sec ±1.71% (91 runs sampled)
  // putChar(0, 97)   x 52,257,064 ops/sec ±1.18% (93 runs sampled)
  // putChar("a")     x 37,559,642 ops/sec ±1.52% (94 runs sampled)
  // putChar(97)      x 40,511,953 ops/sec ±0.87% (96 runs sampled)
